### DISCUSSION TITLE
The unavailable state and numeric_state triggers

### DISCUSSION THREAD

**Discussion by amelchio** [home-assistant/architecture#668] [General] [answered]
## Context

A `numeric_state` trigger will currently fire when its entity changes state from `unavailable` to a numeric value matching the trigger criteria. For example, consider this automation:

```yaml
trigger:
  - platform: numeric_state
    entity_id: sensor.dryer_watt
    below: 10
action:
  - service: script.announce_laundry
```

If the `sensor.dryer_watt` state is `0` and then for some reason changes to `unavailable` and back to `0`, we will get an announcement – even though the dryer most likely was not used during that time.

A simple example where this is unexpected (and undesired) is a sensor that reboots, maybe due to a firmware update.

## Proposal

The `numeric_state` trigger will only fire when its `from` and `to` states are both valid numbers (and a threshold is crossed, of course).

## Consequences

* Numeric state triggers will no longer fire at random.
* If one desires an action on a state change from `unavailable`, it must be handled explicitly. For example, a `state` trigger and a `numeric_state` condition can emulate the current behavior.

**Comment by Adminiuga** [comment]
why not to have a `numeric_state` option, something like `ignore_unavailable_transition` ? then could leave the default to the old behavior thus making a non breaking change.

**Reply by amelchio** [reply]
I would characterize the current behavior as a bug so I don't think backwards compatibility is desired. I think changing the default would _fix_ most automations, not break them.

Most users, I assume, do not expect the current behavior and only find out when their automations suddenly blast in the middle of the night. They do not expect it because it is unintuitive. It is also counter to the documentation which stresses in bold that the trigger only fires when a threshold is crossed.

(Before someone suggests "maybe it is a documentation problem and we should add a note": I don't think it is a documentation problem, the documentation is fine)

It would be interesting to hear some situations where the current behavior is desirable.

**Reply by frenck** [reply]
I do agree with the suggestion, but it does have a side-effect, in case the device is polling or reports e.g., every 2 minutes or something. I do think this proposal makes that extra complex.

```yaml
trigger:
  - platform: homeassistant
    event: start
  - platform: state
    entity_id: 
      - sensor.temperature_outside
      - sensor.temperature_inside
condition:
  - platform: numeric_state
    entity_id: sensor.temperature_outside
    above: sensor.temperature_inside
action:
  - ... # Close window, do heating, HVAC controls, whatever
```

☝️ With the suggestion from this architectural proposal, this won't behave as expected.
If `sensor.temperature_outside` hasn't reported in before Home Assistant startup, it won't fire on the first time it reports in. It needs to report twice (at least) in that case.
Additionally, the proposal only takes `entity_id` account, while `above` referencing the entity will behave differently.

In the end, we don't know the reason why it was unavailable or not. We only know it now matches... it maybe should have triggered or not?

**Reply by frenck** [reply]
Thinking about that, condition-wise, that example might be invalid even.

So a simpler example is:

```yaml
trigger:
  - platform: numeric_state
    entity_id: sensor.temperature_inside
    above: 20
condition:
action:
  - ... # Close window, do cooling, HVAC controls, whatever
```

So, if during start up `sensor.temperature_inside` doesn't get initialized (as device is not yet available) and becomes active later, it needs to provide at least 2 updates before it triggers?

**Reply by frenck** [reply]
And wouldn't this still induce the same issue?

```yaml
trigger:
  - platform: state
    entity_id: 
      - sensor.dryer_watt
      - ...
  ... # whatever more

condition:
  - platform: numeric_state
    entity_id: sensor.temperature_outside
    below: 10

action:
  - service: script.announce_laundry
```

And thus cause inconsistent behavior between the trigger and condition?

**Reply by amelchio** [reply]
I'll address this one first:

```yaml
trigger:
  - platform: numeric_state
    entity_id: sensor.temperature_inside
    above: 20
condition:
action:
  - ... # Close window, do cooling, HVAC controls, whatever
```

> So, if during start up `sensor.temperature_inside` doesn't get initialized (as device is not yet available) and becomes active later, it needs to provide at least 2 updates before it triggers?

With this proposal, it will not trigger at all during initialization.

That is, `[setup] -> unavailable -> 21`: no trigger.

This matches with entities that can provide the initial value immediately, `[setup] -> 21`: no trigger.

It will only trigger when actually crossing the threshold, like `17 -> 21`.

**Reply by amelchio** [reply]
@frenck How does the above sound to you?

**Reply by frenck** [reply]
I think it will lead to just as much unexpected behavior as it tried to solve 🤷

**Reply by amelchio** [reply]
Okay. I cannot really think of any such example. I think "it didn't trigger because it was unavailable" is fairly straightforward behavior.

**Reply by frenck** [reply]
It did trigger because it became available and matches the test now, is fairly straight forward too.

One doesn't know why it was unavailable, and values might have actually changed during unavailability. Therefore missing the trigger due to the suggested proposal in this discussion.

In the current situation, one could filter if they want (use the trigger from/to variables), in the proposed solution it is not possible to filter on such changes, as it is never triggered.

Don't get me wrong. I'm in the middle. Both have pros and cons; However, I do not believe it solves anything, as it introduces a different behavior that doesn't have to be wished for. It additionally kills triggers people could have relied on. 

So, for that reason, I don't believe this suggestion is an improvement.

**Reply by amelchio** [reply]
> It did trigger because it became available and matches the test now, is fairly straight forward too.

That is certainly succint but I am not sure it is straightforward. There are four or five gotchas that need additional notes before it is fully described.

> One doesn't know why it was unavailable, and values might have actually changed during unavailability. Therefore missing the trigger due to the suggested proposal in this discussion.

Heh, that is also my reason for removing the trigger: we don't know what happended. Triggering because something might have changed zero, one or more times at some unknown time in the past does not seem generally useful.

> In the current situation, one could filter if they want (use the trigger from/to variables), in the proposed solution it is not possible to filter on such changes, as it is never triggered.

Well, `unavailable` is not numeric so I think it is suitable that the `numeric_state` trigger does not trigger. One can still match the event using the `state` trigger and that would actually trigger at the right time, not play catch-up.

> Don't get me wrong. I'm in the middle. Both have pros and cons; However, I do not believe it solves anything, as it introduces a different behavior that doesn't have to be wished for. It additionally kills triggers people could have relied on.

There are pros and cons but I do not see an equal amount in each behavior. The only thing that the current way has going for it seems to be that it is the status quo (and I do agree that this has some inherent merit). It would still be interesting to see an automation where the current behavior actually is a good fit.

**Comment by balloob** [comment]
Yeah, I think that this makes sense

**Reply by elupus** [reply]
I think considering the comparison invalid if one is unavailable would work in both @frenck and original proposals example.

In the original arch, the trigger won't fire since you do an below comparison.

In frencks example the trigger would trigger, but the condition would be invalid until both values are available.

**Reply by elupus** [reply]
Ps. This is similar to my issue with templates and unavailable data. Any arithmics on unknown/unavailable data should render result unknown.

**Reply by amelchio** [reply]
All right. Discussion seems to have petered out (not that there ever was much). Since the most negative feedback is "in the middle", I am going to submit a PR for this proposal as soon as my time permits. Thank you.

**Reply by amelchio** [reply]
In [the PR](https://github.com/home-assistant/core/pull/71374), @frenck suggested that this should be discussed some more.

So I tried to find all arguments presented against this proposal and address each of them (in chronological order). I put each one in a separate answer since the discussion format seems unsuited to discuss several points at once.

Please add unrelated points in separate answers as well so we don't miss anything.

**Comment by amelchio** [comment]
frenck [said](https://github.com/home-assistant/architecture/discussions/668#discussioncomment-2149189):

> cause inconsistent behavior between the trigger and condition?

**Reply by amelchio** [reply]
I don't see any inherent consistency problem with the proposal but it works in an inconsistent area.

For example, the condition `"unavailable" < 42` is false but `"unavailable" > 41` is not true.

IMHO the proper way to fix all that is to say that `"unavailable" < 42` is not true or false, but invalid. I think that was also what @elupus alluded to above. 

(It could be a mild form of invalid that would abort an automation but not log an error – that is about the same user experience as returning `False` but it is consistent).

**Reply by frenck** [reply]
You are missing the point, a condition always evaluates, regardless of the trigger. However, with the suggested change we no longer trigger in all the same cases the condition does evaluate (e.g., when becoming available again).

**Reply by amelchio** [reply]
I believe that I do understand how things currently work but I am offering an alternative.

Of course, if one thinks that quietly accepting type errors in the condition is not a wart then such a restructuring will never seem worth the effort.

**Comment by amelchio** [comment]
frenck [said](https://github.com/home-assistant/architecture/discussions/668#discussioncomment-2184039)

> However, I do not believe it solves anything, as it introduces a different behavior that doesn't have to be wished for.

**Reply by amelchio** [reply]
We can say that about any change but it sounds like a logical fallacy. A change can solve something even if maybe somebody doesn't like the change.

**Reply by frenck** [reply]
Sure, makes another use case harder to use, even more, it introduces a breaking change without a real problem to solve IMHO. In that case, we should just NOT do it.

**Reply by amelchio** [reply]
I explained the problem in the OP. It may not be our biggest problem but it is a real problem that I have experienced myself and that has caused forum posts.

Unavailability is just hard and IMHO we are causing some confusion by pretending that we can handle it automatically.

**Comment by amelchio** [comment]
frenck [said](https://github.com/home-assistant/architecture/discussions/668#discussioncomment-2184039)

> It additionally kills triggers people could have relied on.

**Reply by amelchio** [reply]
We often remove things that people rely on. Not too lightly of course but an incompatibility is a breaking change, not a showstopper.

In this case I think the proposal is a usability improvement that is worth the breaking change in a corner case.

**Reply by frenck** [reply]
See my previous response, but that is hard to track in the way this thread is now set up... not sure why it is needed to create separate sub-threads for each part of my responses 🤷

**Reply by amelchio** [reply]
Yeah, I can't say that I love the split myself. AFAIK we moved from GitHub issues to GitHub discussions because a long list of comments is incomprehensible. It very easily ends up with ideas or whole posts being missed or ignored and I wanted to avoid that.

But it seems that GitHub is just not suited for a long discussion with multiple subthreads. I guess having no way to reply to an individual comment is a big part of that problem since every time a discussion branches it gets hard to follow.

**Comment by amelchio** [comment]
frenck [said](https://github.com/home-assistant/core/pull/71374#issuecomment-1120147623):

> This will assume that if a device is offline and comes back, it was in the range during it being offline.

**Reply by amelchio** [reply]
There is no such assumption. We skip the firing because we have no knowledge that something happended. Not because we do know that nothing happended.

It seems comparable to a Home Assistant restart to me. A watched entity might indeed change its value while a restart goes on. We don't track that and we don't fire triggers for initial states. An unavailable entity is just a tiny version of that.

**Reply by frenck** [reply]
> We skip the firing because we have no knowledge that something happened.

Exactly, so we should fire. As the only thing we know, it is now in the range after coming back.

**Comment by amelchio** [comment]
frenck [said](https://github.com/home-assistant/core/pull/71374#issuecomment-1120950885):

> IMHO, this behavior is inconsistent compared to all other triggers

**Reply by amelchio** [reply]
The `numeric_state` trigger is special because it can see invalid values (i.e. non-numbers). I believe that no other trigger has to consider an invalid input. So I think it is a unique situation, not inconsistency.

Incidentally, this proposal actually *removes* an inconsistency: no other trigger fires because something *might have* happended.

**Reply by frenck** [reply]
I don't think it's special.

> I believe that no other trigger has to consider an invalid input. 

Not sure what that has to do with that.

**Reply by amelchio** [reply]
Okay, maybe I guessed wrong. Can you elaborate on why you think it is inconsistent compared to all other triggers?

**Reply by frenck** [reply]
> Can you elaborate on why you think it is inconsistent compared to all other triggers

Sure, for example:

```yaml
automation:
  trigger:
    - platform: state
      entity_id: binary_sensor.motion
      to: "on"
```

This will trigger, regardless of if it was unavailable before or not. If it becomes available again and is "on" it will retrigger once again.

Another example:

```yaml
automation:
  trigger:
    - platform: zone
      entity_id: device_tracker.frenck
      zone: zone.home
      event: enter
```

The above trigger will trigger again once the device tracker becomes available again.

**Reply by amelchio** [reply]
Thank you.

For there to be a general rule about retriggering, it seems like this should also fire but it does not:

```yaml
automation:
  trigger:
    - platform: state
      entity_id: binary_sensor.motion
      from: "off"
      to: "on"
```

About zones, I don't normally use those so I may be missing something. But I tried to reproduce the zone example and I failed. It seems that the zone trigger actually bails out when the entity was unavailable and the coordinates thus are missing: https://github.com/home-assistant/core/blob/ebc883b43ff9ca00d3426bbd0bdfdab707576252/homeassistant/components/zone/trigger.py#L76-L77

Anyway, I made this proposal only about `numeric_state` since I knew that a full-on discussion of `unavailable` could explode into something that the GitHub discussion format is unfit to handle. I acknowledge that if there is consensus about this proposal then it warrants a review of other triggers for similar issues.

**Comment by amelchio** [comment]
frenck [said](https://github.com/home-assistant/core/pull/71374#issuecomment-1120950885):

> [the proposal] has the potential to miss triggers (which is not wished for IMHO)

**Reply by amelchio** [reply]
The trigger is already missed today. The trigger is about crossing a threshold. What fires today is not the proper trigger but a surrogate that might fire long after the threshold was actually crossed. Worse, it might fire even though the threshold was never crossed (my original example at the top).

Furthermore, only one surrogate trigger is fired. But the entity might have crossed the threshold multiple times while unavailable. So the current implementation already "has the potential to miss triggers", it is not a flaw that the proposal introduces.

**Reply by frenck** [reply]
> What fires today is not the proper trigger but a surrogate that might fire long after the threshold was actually crossed.

We don't know if it is a surrogate. That is the whole point.

> Worse, it might fire even though the threshold was never crossed (my original example at the top).

That is not worse IMHO. We simply don't know.

> But the entity might have crossed the threshold multiple times while unavailable. So the current implementation already "has the potential to miss triggers",

That is pulling my concerns way out of context to prove a point, and not relevant for this discussion at all IMHO.

**Comment by balloob** [comment]
Here is a variation of both proposals:

Make the numeric state trigger track the last valid state and use that when a new valid state comes in.

Currently a rule matching on `above: 22`:

- 19 -> unavailable -> 22: trigger
- 22 -> unavailable -> 22: trigger
- unavailable -> 22: trigger

With this proposal:

- 19 -> unavailable -> 22: trigger
- 22 -> unavailable -> 22: not trigger
- unavailable -> 22: trigger

**Reply by amelchio** [reply]
What about 24 -> unavailable -> 23 ?

**Reply by frenck** [reply]
Doesn't trigger, as it is still above the condition (above 22, it has triggered before).

**Reply by amelchio** [reply]
Unfortunately, all of the concerns raised by frenck still apply with this variant.

**Reply by balloob** [reply]
It's fine, this was pre-approved 👍

**Comment by amelchio** [comment, marked as answer]
Proposal withdrawn.

**Reply by melyux** [reply]
Why? Ran into this bug today.

**Reply by amelchio** [reply]
Right, the bug still exists and the proposal is still valid.

The withdrawal was more about me leaving this process behind.
